include/swift/Runtime/Concurrent.h

```diff
@@ -527,7 +527,7 @@ template <class ElemTy> struct ConcurrentReadableArray {
       
       storage = newStorage;
       Capacity = newCapacity;
-      Elements.store(storage, std::memory_order_release);
+      Elements.store(storage, std::memory_order_seq_cst);
```

Comment:
I think the fix is correct but the reasoning doesn't feel right to me at all. The problem isn't that the readers are picking up the old value but that the writer isn't reasoning correctly about whether there are readers or not. After all, this proposed fix is on the writer side of the design.

On x86, the only C++11 memory model hint that matters "sequential consistency" and even then, it only matters for stores. There are two reasons for this: 1) atomic x86 instructions don't support anything other than sequential consistency and 2) the only reordering allowed by x86's "total store order" memory model is newer loads being moved ahead of older stores.

By switching `Elements.store(...)` to sequential consistency, we're forcing `ReaderCount.load()` to happen after the store.

I think if one is going to use `std::memory_order_seq_cst`, then one really ought to document at least one example in a comment of an actual subsequent load in the same thread that must come after the given store.

Reply:
Here is the ASCII art of what I think is happening:
```
Effective program order due to out-of-order loads.

Writer                 | Reader
-----------------------+--------------------
............ ReaderCount == 0 ..............
                       |
 -> ReaderCount.load() |
/                      |
|                      | incrementReaders()
|                      | storage = Elements.load(SWIFT_MEMORY_ORDER_CONSUME);
|                      |
| Elements.store()     |
\                      |
 *ReaderCount.load()   |

Now the writer wrongly thinks there are zero readers of the old data.
```